include/vcmi/scripting/Service.h

```diff
@@ -0,0 +1,80 @@
+/*
+ * scripting/Service.h, part of VCMI engine
+ *
+ * Authors: listed in file AUTHORS in main folder
+ *
+ * License: GNU General Public License v2.0 or later
+ * Full text of license available in license.txt file, in main folder
+ *
+ */
+
+#pragma once
+
+#include <vcmi/Environment.h>
+
+class Services;
+class JsonNode;
+class ServerCallback;
+
+namespace scripting
+{
+
+using BattleCb = ::Environment::BattleCb;
+using GameCb = ::Environment::GameCb;
+
+class DLL_LINKAGE Context
+{
+public:
+	virtual ~Context() = default;
+
+	virtual void run(const JsonNode & initialState) = 0;
+	virtual void run(ServerCallback * server, const JsonNode & initialState) = 0;
+
+	virtual JsonNode callGlobal(const std::string & name, const JsonNode & parameters) = 0;
+	virtual JsonNode callGlobal(ServerCallback * server, const std::string & name, const JsonNode & parameters) = 0;
+
+	virtual void setGlobal(const std::string & name, int value) = 0;
+	virtual void setGlobal(const std::string & name, const std::string & value) = 0;
+	virtual void setGlobal(const std::string & name, double value) = 0;
+	virtual void setGlobal(const std::string & name, const JsonNode & value) = 0;
+
+	virtual void getGlobal(const std::string & name, int & value) = 0;
+	virtual void getGlobal(const std::string & name, std::string & value) = 0;
+	virtual void getGlobal(const std::string & name, double & value) = 0;
+	virtual void getGlobal(const std::string & name, JsonNode & value) = 0;
+
+	virtual JsonNode saveState() = 0;
+};
+
+class DLL_LINKAGE Script
+{
+public:
+	virtual ~Script() = default;
+
+	virtual const std::string & getName() const = 0;
+	virtual const std::string & getSource() const = 0;
+
+	virtual std::shared_ptr<Context> createContext(const Environment * env) const = 0;
+};
+
+class DLL_LINKAGE Pool
```

Comment:
Not it do not cache unused contexts but it could (and should be) be improved in that direction for BattleAI at least, for now it just ensures that script uses correct execution context. By design all scripts are independent and each runs on it own context.